### ISSUE TITLE
Bazel build failure

### ISSUE THREAD

**Issue by childss** [krpc/krpc#410]
I've been trying to get setup to build the project with the hopes of contributing in some way (or at least playing with implementing a new language client), but haven't succeeded yet. After poking around the docs and Github issues, it seems like OSX might not be as well supported so I decided to spin up a brand new Ubuntu AWS box and see if I could get it working there. I installed all the base system dependencies as recommended in the "Compiling kRPC" docs but when I run `bazel build //:krpc` I get a failure apparently related to Sphinx:

```
$ bazel build //:krpc --verbose_failures
............................
INFO: Analysed target //:krpc (62 packages loaded).
INFO: Found 1 target...
ERROR: /home/ubuntu/krpc/doc/BUILD:46:1: Setting up python script sphinx-build failed (Exit 1): py_script-setup-sphinx-build failed: error executing command bazel-out/host/genfiles/doc/py_script-setup-sphinx-buil
d

Use --sandbox_debug to see verbose messages from the sandbox
Command "/home/ubuntu/.cache/bazel/_bazel_ubuntu/8c03bc02ce2a81e3fdca16cf2591e999/bazel-sandbox/8546816907699913227/execroot/krpc/bazel-out/host/genfiles/doc/py_script-env-sphinx-build.tmp-create-py-env.2/bin/pyt
hon -u -c "import setuptools, tokenize;__file__='/tmp/pip-ciurwI-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" in
stall --record /tmp/pip-lxSjh2-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/ubuntu/.cache/bazel/_bazel_ubuntu/8c03bc02ce2a81e3fdca16cf2591e999/bazel-sandbox/8546
816907699913227/execroot/krpc/bazel-out/host/genfiles/doc/py_script-env-sphinx-build.tmp-create-py-env.2/include/site/python2.7/lxml" failed with error code 1 in /tmp/pip-ciurwI-build/
Target //:krpc failed to build
INFO: Elapsed time: 60.398s, Critical Path: 22.16s
FAILED: Build did NOT complete successfully
```

Given that the error message implicates Sphinx, I also tried installing that on the system but to no avail. I noticed that there's a Docker image/scripts for running on Travis (including a handy-looking `local.sh` script) so I gave that a shot but it ended in a similar failure:

```
$ ./tools/travis-ci/local.sh
... snip ...
[444 / 445] Still waiting for 1 job to complete:
      Running (standalone):
        Setting up python script sphinx-build, 10 s
ERROR: /build/krpc/doc/BUILD:46:1: Setting up python script sphinx-build failed: py_script-setup-sphinx-build failed: error executing command
  (cd /build/.cache/bazel/_bazel_build/d75c572c8955edc6065ad0155a0d4696/execroot/krpc && \
  exec env - \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
  bazel-out/local-fastbuild/genfiles/doc/py_script-setup-sphinx-build): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
Command "/build/.cache/bazel/_bazel_build/d75c572c8955edc6065ad0155a0d4696/execroot/krpc/bazel-out/local-fastbuild/genfiles/doc/py_script-env-sphinx-build.tmp-create-py-env.5647/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-dGFsy6-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-cz16zt-record/install-record.txt --single-version-externally-managed --compile --install-headers /build/.cache/bazel/_bazel_build/d75c572c8955edc6065ad0155a0d4696/execroot/krpc/bazel-out/local-fastbuild/genfiles/doc/py_script-env-sphinx-build.tmp-create-py-env.5647/include/site/python2.7/lxml" failed with error code 1 in /tmp/pip-dGFsy6-build/
INFO: Building complete.
INFO: Elapsed time: 193.969s, Critical Path: 56.99s
```

Anyway, I'm hoping that this will trigger a memory for you when you were setting up the build scripts or be able to give me some pointers on ways to debug this as I'm unfamiliar with Bazel and the sandbox tool. Thanks!

**Comment by djungelorm** [comment]
I think someone did get it compiling under OSX a while ago, so might work, but as you say it is possible that it won't work.

Installing sphinx on the system shouldn't affect anything. The build scripts use virtualenv and pip to install sphinx into the sandboxed environment.

The build step that is failing is the one that creates this environment. The code for this custom build step is mostly in `tools/build/python.bzl`. You should be able to get more details as to what went wrong by removing the --quiet flag from being passed to pip when it's run, i.e. this patch: 
https://gist.github.com/djungelorm/7e841a0791573b6abf5f7962e58a0a2f
That should give more info as to what is going wrong.

(PS. I'm going to be away on holiday over the next two weeks so may be slow to respond!)

**Comment by childss** [comment]
Well, that at least gets me a bit more to grasp for further debugging.

```
Successfully installed Jinja2-2.8
Processing ./external/python_lxml/file/lxml-3.6.4.tar.gz
Installing collected packages: lxml
  Running setup.py install for lxml: started
The directory '/home/ubuntu/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/home/ubuntu/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.

(and quite a few more duplicates of these lines before failing)
```

Naturally I've checked ownership and permissions of these directories and they look fine. I did some googling and found some things about problems installing lxml in virtualenv but it doesn't seem to be my problem (also `pip install lxml` works both in and out of a virtualenv). I'll try some more poking around in the bazel files.

Enjoy your holiday!

**Comment by djungelorm** [comment]
I don't think the permissions are a problem - when sandboxed, pip can't see that directory so the error is actually because it thinks it doesn't exist. This only prevents it caching downloads though, which we don't really need.

Are you running with bazel build //:krpc --verbose_failures? Some of the log might be hidden from you if not!

**Comment by childss** [comment]
I finally figured it out! I just wanted to see if I could get it working on Ubuntu and didn't want to futz with setting up a VM host on my local machine so I booted a cheap AWS instance... with 512 MB of memory. Turns out that's not enough to compile libxml as revealed by `dmesg | tail` :)

I got it to work after activating a 512 MB swapfile. Looks like it'll be worth another try on OSX or at least a beefier AWS instance. Thanks again for the help; the idea to remove `--quiet` really gave the boost necessary to figure this out!
